blockprod/src/detail/mod.rs

```diff
@@ -306,33 +310,95 @@ impl BlockProduction {
                 });
             }
 
-            tokio::select! {
-                _ = &mut cancel_receiver => {
-                    // TODO: test cancellations
-                    stop_flag.store(true, std::sync::atomic::Ordering::SeqCst);
-                    return Err(BlockProductionError::Cancelled);
+            'consume_chainstate_events: loop {
+                tokio::select! {
+                    _ = chainstate_receiver.changed() => {
+                        // TODO: test broadcasts
+                        self.broadcast_new_tip(*chainstate_receiver.borrow());
+                        continue 'consume_chainstate_events;
+                    }
```

Comment:
This is not really the appropriate place for this. Receiving chainstate events shouldn't be done in generate_block, but in a possible event loop related to block production itself.

So what you're doing is: ChainstateEvent -> generate block -> broadcast_new_tip -> generate_block -> stop
What should be done is: ChainstateEvent -> BlockProduction -> broadcast_new_tip -> generate_block -> stop

There are many reasons for why this should be done like this, including circularity of the current logic. I think the confusion here is that you need an event loop... well... start a new task that receives chainstate events! That's how you create an event loop.

So once you connect chainstate's new tip with the function broadcast_new_tip, your job is basically done, because generate_block already has a cancel_receiver that solves that problem to you. Notice that part of the design flaw here is that you have to get through the `select!` macro twice.

Reply:
Updated to move into its own `run()` function called by the node runner